langchain/vectorstores/matching_engine.py

```diff
@@ -0,0 +1,373 @@
+"""Vertex Matching Engine implementation of the vector store."""
+
+import json
+import logging
+import time
+import uuid
+
+from google.cloud import aiplatform
+from google.cloud import storage
+from google.oauth2 import service_account
+from typing import Any, Iterable, List, Optional, Type, Union
+
+from langchain.docstore.document import Document
+from langchain.embeddings import TensorflowHubEmbeddings
+from langchain.embeddings.base import Embeddings
+from langchain.vectorstores.base import VectorStore
+
+logger = logging.getLogger()
+
+HUB_MODEL = "https://tfhub.dev/google/universal-sentence-encoder-multilingual/3"
+
+
+class MatchingEngine(VectorStore):
+    """Vertex Matching Engine implementation of the vector store.
+
+    While the embeddings are stored in the Matching Engine, the embedded
+    documents will be stored in GCS.
+
+    Note that this implementation is mostly meant for reading if you are
+    planning to do a real time implementation. While reading is a real time
+    operation, updating the index takes close to one hour."""
+
+    def __init__(self,
+        project_id: str,
```

Comment:
we're generally trying to move towards constructor only taking as input and storing repeatedly used objects, and everything that's only needed for initialization should be done outside constructor. and then if we want we can still have a convenience class method named something like `from_components` that puts together various id's and such into clients and indices and endpoints.

so in this case maybe that'd be somthing like
```python
def __init__(self, index, endpoint, client):
  ...
  
@classmethod
def from_components(project_id, region, gcs_bucket_uri, index_id, ...):
  index = _create_index_by_id(index_id)
  endpoint = ...
  client = ...
  return cls(index, endpoint, ...)
```

thoughts?